src/transformStateWithClones.js

```diff
@@ -7,7 +7,45 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = addProperties(currentState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        currentState = removeProperties(currentState, action.keysToRemove);
+        break;
+
+      case 'clear':
+        currentState = clearProperties();
+        break;
+    }
+    stateHistory.push(currentState);
```

Comment:
[DONT REPEAT YOURSELF] - Instead of pushing the `currentState` to the `stateHistory` array inside the `switch` statement, it's better to do it just once afterwards. This reduces repetition and makes the code cleaner.